src/neuroagent/scripts/avalidate_tool_calls.py

```diff
@@ -0,0 +1,251 @@
+    optional_tools: list[str],
+    forbidden_tools: list[str],
+) -> tuple[bool, str]:
+    """
+    Validate the sequence of tool calls against required, optional, and forbidden tools.
+
+    Args:
+    ----
+        required_tools (List): A list of tools that must be called in the specified order.
+        actual_tool_calls (List): A list of tools that were actually called.
+        optional_tools (List): A list of tools that can be called but are not required.
+        forbidden_tools (List): A list of tools that must not be called.
+
+    Returns
+    -------
+        tuple: A tuple containing a boolean and a string message. The boolean is True if the
+               validation is successful, otherwise False. The string message provides details
+               about the validation result.
+    """
+    # Check for forbidden tools
+    if inter := set(actual_tool_calls) & set(forbidden_tools):
+        return False, f"Forbidden tool(s) called: {inter}"
+
+    # Validate required tools order
+    order = 0
+    for tool in actual_tool_calls:
+        if order < len(required_tools) and tool == required_tools[order]:
+            order += 1
+        elif tool in optional_tools or (
+            order > 0 and tool == required_tools[order - 1]
+        ):
+            continue
+        elif tool not in required_tools[:order]:
+            return False, f"Unexpected tool called: {tool}"
+
+    # Check if all required tools were called
+    if order != len(required_tools):
+        return False, "Not all required tools were called"
+
+    return True, "All required tools called correctly"
+
+
+def main() -> None:
+    """
+    Execute the tool call validation process.
+
+    This function sets up the argument parser to handle command-line arguments,
+    specifically for specifying the base URL, port, data file path, and output
+    CSV file name. It then calls the validate_tool_calls_async function with
+    the provided arguments to perform the validation of tool calls and save
+    the results.
+
+    The function is designed to be the entry point when the script is run
+    directly from the command line.
+    """
+    parser = argparse.ArgumentParser(
+        description="Run tool call tests and save results."
+    )
+    parser.add_argument(
+        "--base",
+        type=str,
+        default="localhost",
+        help="Base URL for the API",
+    )
+    parser.add_argument(
+        "--port",
+        type=int,
+        default=8000,
+        help="Port number for the API",
+    )
+    parser.add_argument(
+        "--data",
+        type=str,
+        default="tests/data/tool_calls.json",
+        help="Path to the JSON file containing test case data",
+    )
+    parser.add_argument(
+        "--output",
+        type=str,
+        default="tool_call_evaluation.csv",
+        help="Output CSV file for results",
+    )
+    args = parser.parse_args()
+
+    # Construct the full base URL
+    base_url = f"http://{args.base}:{args.port}"
```

Comment:
This would fail if we point to aws for instance. I suggest we let the user include the protocol in the `base_url`